packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart

```diff
@@ -0,0 +1,823 @@
+    // applied with dart io and custom parsing.
+    final File appManifestFile = androidDir
+      .childDirectory('app')
+      .childDirectory('src')
+      .childDirectory('main')
+      .childFile('AndroidManifest.xml');
+    _inputs.add(appManifestFile);
+    XmlDocument document;
+    try {
+      document = XmlDocument.parse(appManifestFile.readAsStringSync());
+    } on XmlParserException {
+      throwToolExit('Error parsing $appManifestFile '
+                    'Please ensure that the android manifest is a valid XML document and try again.');
+    } on FileSystemException {
+      throwToolExit('Error reading $appManifestFile even though it exists. '
+                    'Please ensure that you have read permission to this file and try again.');
+    }
+    // Create loading unit mapping.
+    final Map<int, String> mapping = <int, String>{};
+    for (final DeferredComponent component in components) {
+      component.assignLoadingUnits(generatedLoadingUnits);
+      for (final LoadingUnit unit in component.loadingUnits) {
+        if (!mapping.containsKey(unit.id)) {
+          mapping[unit.id] = component.name;
+        }
+      }
+    }
+    // Encode the mapping as a string.
+    String encodedMapping = '';
+    for (final int key in mapping.keys) {
+      encodedMapping += '$key:${mapping[key]},';
+    }
+    encodedMapping = encodedMapping.substring(0, encodedMapping.length - 1); // remove trailing comma.
+    // Check for existing metadata entry and see if needs changes.
+    bool exists = false;
+    bool needsChange = false;
+    const String mappingKey = 'io.flutter.embedding.engine.deferredcomponents.DeferredComponentManager.loadingUnitMapping';
+    for (final XmlElement metaData in document.findAllElements('meta-data')) {
+      final String name = metaData.getAttribute('android:name');
+      if (name == mappingKey) {
+        exists = true;
+        final String storedMappingString = metaData.getAttribute('android:value');
+        if (storedMappingString != encodedMapping) {
+          needsChange = true;
+        }
+      }
+    }
+    if (needsChange || !exists) {
+      lines = appManifestFile.readAsLinesSync();
+      List<String> cleanedLines = <String>[];
+      if (exists) {
+        // Remove existing mapping entry. We do this line by line
+        // to maintain the formatting and comments in the file.
+        for (int i = 0; i < lines.length; i++) {
+          final String line = lines[i];
+          if (line.trim().startsWith(RegExp(r'<meta-data'))) {
+            // Reconstruct metadata entry regardless of how many lines it is split over
+            String metaDataLine = '';
+            int j = i;
+            while (!lines[j].contains('/>')) {
+              metaDataLine += lines[j];
+              j++;
+            }
+            metaDataLine += lines[j];
+            if (metaDataLine.contains(RegExp(r'<meta-data[\s\n]*android:name[\s\n]*=[\s\n]*"' + mappingKey + r'"'))) {
```

Comment:
In general RegExps should be hoisted into statics and extensively documented as to what they are intended to match.